src/python/pants/backend/tools/yamllint/rules.py

```diff
@@ -0,0 +1,185 @@
+    while True:
+        candidate_config_file_path = os.path.join(dir, config_file)
+        if candidate_config_file_path in files:
+            return candidate_config_file_path
+
+        if dir == "":
+            return None
+        dir = os.path.dirname(dir)
+
+
+@rule
+async def gather_config_files(
+    request: YamllintConfigFilesRequest, yamllint: Yamllint
+) -> YamllintConfigFiles:
+    """Gather yamllint configuration files."""
+    source_dirs = frozenset(os.path.dirname(path) for path in request.filepaths)
+    source_dirs_with_ancestors = {"", *source_dirs}
+    for source_dir in source_dirs:
+        source_dir_parts = source_dir.split(os.path.sep)
+        source_dir_parts.pop()
+        while source_dir_parts:
+            source_dirs_with_ancestors.add(os.path.sep.join(source_dir_parts))
+            source_dir_parts.pop()
+
+    config_file_globs = [
+        os.path.join(dir, yamllint.config_file_name) for dir in source_dirs_with_ancestors
+    ]
+    config_files_snapshot = await Get(Snapshot, PathGlobs(config_file_globs))
+    config_files_set = set(config_files_snapshot.files)
+
+    source_dir_to_config_file: dict[str, str] = {}
+    for source_dir in source_dirs:
+        config_file = find_nearest_ancestor_file(
+            config_files_set, source_dir, yamllint.config_file_name
+        )
+        if config_file:
+            source_dir_to_config_file[source_dir] = config_file
+
+    return YamllintConfigFiles(config_files_snapshot, FrozenDict(source_dir_to_config_file))
+
+
+@rule
+async def partition_inputs(
+    request: YamllintRequest.PartitionRequest, yamllint: Yamllint
+) -> Partitions[Any, PartitionInfo]:
+    if yamllint.skip:
+        return Partitions()
+
+    matching_filepaths = FilespecMatcher(
+        includes=yamllint.file_glob_include, excludes=yamllint.file_glob_exclude
+    ).matches(request.files)
+
+    config_files = await Get(
+        YamllintConfigFiles, YamllintConfigFilesRequest(filepaths=tuple(matching_filepaths))
+    )
+
+    default_source_files: set[str] = set()
+    source_files_by_config_file: dict[str, set[str]] = defaultdict(set)
+    for source_dir, files_in_source_dir in group_by_dir(matching_filepaths).items():
+        files = (os.path.join(source_dir, name) for name in files_in_source_dir)
+        if source_dir in config_files.source_dir_to_config_files:
+            config_file = config_files.source_dir_to_config_files[source_dir]
+            source_files_by_config_file[config_file].update(files)
+        else:
+            default_source_files.update(files)
+
+    config_file_snapshots = await MultiGet(
+        Get(Snapshot, DigestSubset(config_files.snapshot.digest, PathGlobs([config_file])))
+        for config_file in source_files_by_config_file
+    )
+
+    return Partitions(
+        (
+            *(
+                Partition(tuple(files), PartitionInfo(config_snapshot=config_snapshot))
```

Comment:
`set`s change order from invocation to invocation, and since you put this in a tuple (ordered) this will result in cache misses for the same contents.

`sorted` is your friend.

(same below)